typeql/modules/ROOT/pages/pipelines/put.adoc

```diff
@@ -1,17 +1,83 @@
-// ----
-// match
-//   $person-1 isa person, has username "<username 1>";
-//   $person-2 isa person, has username "<username 2>";
-// put
-//   friendship (friend: $person-1, friend: $person-2);
-// ----
+A put stage will xref:{page-version}@typeql::pipelines/insert.adoc[insert] data according to a pattern if no existing data xref:{page-version}@typeql::pipelines/match.adoc[matches] that pattern.
+
+== Syntax
+
+As a put stage may insert data, the body of a put stage is a sequence of valid xref:{page-version}@typeql::pipelines/insert.adoc[insert] statements.
+
+[,typeql]
+----
+put <insert-statement> [ <insert-statement> ... ]
+----
+
+== Example
+
+The following query creates a friendship relation between two persons, if one does not already exist.
+
+[,typeql]
+----
+match
+  $person-1 isa person, has username "Alice";
+  $person-2 isa person, has username "Bob";
+put
+  friendship (friend: $person-1, friend: $person-2);
+----
+
+== Behaviour
+
+The behaviour of a `put` stage can be described in pseudo-code as:
+[,python]
+----
+def run_put(pattern, input_rows):
+    output_rows = []
+    for row in input_rows:
+        matched_rows = run_match(pattern, row)
+        if matched_rows.is_empty():
+            inserted_row = run_insert(pattern, row)
+            output_rows += [inserted_row]
+        else:
+            output_rows += matched_rows
+    return output_rows
+----
+
+Hence, a put stage is expected to be only as performant as running both the match and the insert stage.
+
+=== No partial matching
+The put matches the entire pattern or inserts the entire pattern. Partial matching and partial inserts are not performed.
+
+To illustrate, consider running the following queries sequentially.
+
+[,typeql]
+----
+# Creates some initial data
+put $p isa person, has name "Alice";
+
+# Does nothing, as the match succeeds
+put $p1 isa person, has name "Alice";
+
+# Creates a new person with name "Bob"
+put $p2 isa person, has name "Bob";
+
+# Creates a new person with name "Alice" & age 10.
+put $p3 isa person, has name "Alice", has age 10;
+----
+The statement `put $p3...` creates a new person because the **whole** pattern fails to match,
+as there is no existing person with name "Alice" **AND** age 10.
+
+To achieve the desired result of not re-creating the person if only the age attribute is missing,
+`put` stages can be pipelined.
+
+[,typeql]
+----
+put $p4 isa person, has name "Bob"; # Matches the person with name "Bob", since it exists
+put $p4 isa person, has age 11;     # Adds an attribute to the same person
+----
+
+=== Isolation from concurrent transactions
+Though put stages are meant to create data patterns which don't exist in the database,
+they still run under the xref:{page-version}@manual::queries/transactions.adoc[snapshot-isolation] of TypeDB transactions.
+This means they will insert data even if a concurrent transaction has inserted data which satisfies the match - leading to duplication if both transactions commit.
```

Comment:
I would add - to prevent this, the recommended pattern is to ensure the `put` pattern will insert data corresponding to @card or @key constraints, which will allow only one concurrent and overlapping `put` to succed.

Reply:
Doesn't that only flag the edges which have the `@card` or `@key` ?

Reply:
But because put is all-or-nothing that would prevent the put from running more than once across transactions which is what the user would want right?

Reply:
ie. any conflict within the put block would make the whole thing at-most-once

Reply:
Updated. Please check.